db/compaction/compaction.cc

```diff
@@ -563,6 +563,47 @@ bool Compaction::KeyNotExistsBeyondOutputLevel(
   return false;
 }
 
+bool Compaction::KeyRangeNotExistsBeyondOutputLevel(
+    const Slice& begin_key, const Slice& end_key,
+    std::vector<size_t>* level_ptrs) const {
+  assert(input_version_ != nullptr);
+  assert(level_ptrs != nullptr);
+  assert(level_ptrs->size() == static_cast<size_t>(number_levels_));
+  assert(cfd_->user_comparator()->CompareWithoutTimestamp(begin_key, end_key) <
+         0);
+  if (bottommost_level_) {
+    return true /* does not overlap */;
+  } else if (output_level_ != 0 &&
+             cfd_->ioptions()->compaction_style == kCompactionStyleLevel) {
+    const Comparator* user_cmp = cfd_->user_comparator();
+    for (int lvl = output_level_ + 1; lvl < number_levels_; lvl++) {
+      const std::vector<FileMetaData*>& files =
+          input_vstorage_->LevelFiles(lvl);
+      for (; level_ptrs->at(lvl) < files.size(); level_ptrs->at(lvl)++) {
+        auto* f = files[level_ptrs->at(lvl)];
+        // Advanced until the first file with begin_key <= f->largest.user_key()
+        if (user_cmp->CompareWithoutTimestamp(begin_key,
+                                              f->largest.user_key()) > 0) {
+          continue;
+        }
+        // We know that the previous file prev_f, if exists, has
+        // prev_f->largest.user_key() < begin_key.
+        if (user_cmp->CompareWithoutTimestamp(end_key,
+                                              f->smallest.user_key()) <= 0) {
+          // not overlapping with this level
+          break;
```

Comment:
Concluding non-overlap in the equality case feels risky when considering this is called with the "improperly truncated" range tombstones, particularly `tombstone_end`. Are we sure the read path will do the right thing?

Reply:
I thought about this too and I think so. When `tombstone_end` is some k@s where s is not kMaxSequenceNumber (I assume this is "improperly truncated"), the range tombstone does not cover any key k with seqno < s. All keys beyond the output level with user key equals to k should have sequence number < s, so they are not covered by this range tombstone and should be safe to drop it.

Reply:
Yes that is what I meant - I reread https://github.com/cockroachdb/pebble/blob/master/docs/range_deletions.md basically every time I review a DeleteRange change, which is where this terminology originates.

OK, I guess the range tombstone fragment in the next file should cover any key k with seqno < s.

Reply:
Thanks for the review! I end up needing to read our range deletion implementations every time I make a non-trivial change too.